src/components/MessageInput/WithDragAndDropUpload.tsx

```diff
@@ -0,0 +1,149 @@
+import React, {
+  CSSProperties,
+  ElementType,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+import {
+  MessageInputContextValue,
+  useChannelStateContext,
+  useMessageInputContext,
+  useTranslationContext,
+} from '../../context';
+import { useDropzone } from 'react-dropzone';
+import clsx from 'clsx';
+
+const DragAndDropUploadContext = React.createContext<{
+  fileQueue: File[];
+  addFilesToQueue: ((files: File[]) => void) | null;
+  removeFilesFromQueue: ((files: File[]) => void) | null;
+}>({
+  addFilesToQueue: null,
+  fileQueue: [],
+  removeFilesFromQueue: null,
+});
+
+export const useDragAndDropUploadContext = () => useContext(DragAndDropUploadContext);
+
+/**
+ * @private To maintain top -> bottom data flow, the drag-and-drop functionality allows dragging any files to the queue - the closest
+ * `MessageInputProvider` will be notified through `DragAndDropUploadContext.fileQueue` and starts the upload with `uploadNewAttachments`,
+ * forwarded files are immediately after removed from the queue.
+ */
+export const useHandleDragAndDropQueuedFiles = ({
+  uploadNewFiles,
+}: MessageInputContextValue) => {
+  const { fileQueue, removeFilesFromQueue } = useDragAndDropUploadContext();
+
+  // const uploadNewFilesReference = useRef<((f: File[]) => void) | undefined>(undefined);
+
+  // uploadNewFilesReference.current = (f) => messageInputContextValue.uploadNewFiles(f);
+
+  useEffect(() => {
+    if (!removeFilesFromQueue) return;
+
+    // uploadNewFilesReference.current?.(fileQueue);
+    uploadNewFiles(fileQueue);
+
+    removeFilesFromQueue(fileQueue);
+  }, [fileQueue, removeFilesFromQueue, uploadNewFiles]);
+};
+
+/**
+ * Wrapper to replace now deprecated `Channel.dragAndDropWindow` option.
+ *
+ * @example
+ * ```tsx
+ * <Channel>
+ *  <WithDragAndDropUpload as="section" className="message-list-dnd-wrapper">
+ *    <Window>
+ *      <MessageList />
+ *      <MessageInput />
+ *    </Window>
+ *  </WithDragAndDropUpload>
+ *  <Thread />
+ * <Channel>
+ * ```
+ */
+export const WithDragAndDropUpload = ({
+  as: Component = 'div',
```

Comment:
Could we change this prop name to a substantive?

Reply:
Changed to prop name to a `component` (as is [Material UI standard](https://mui.com/material-ui/react-box/#basics)).